ui/frontend/editor/MonacoEditor.tsx

```diff
@@ -0,0 +1,86 @@
+import React from 'react';
+import { CommonEditorProps, Position } from '../types';
+import MonacoReact, { Monaco } from "@monaco-editor/react";
+
+import styles from './Editor.module.css';
+import { config, grammar } from './rust_monaco_def';
+
+class CodeByteOffsets {
+  readonly code: string;
+  readonly lines: string[];
+
+  constructor(code: string) {
+    this.code = code;
+    this.lines = code.split('\n');
+  }
+
+  public lineToOffsets(line: number) {
+    const precedingBytes = this.bytesBeforeLine(line);
+
+    const highlightedLine = this.lines[line];
+    const highlightedBytes = highlightedLine.length;
+
+    return [precedingBytes, precedingBytes + highlightedBytes];
+  }
+
+  public rangeToOffsets(start: Position, end: Position) {
+    const startBytes = this.positionToBytes(start);
+    const endBytes = this.positionToBytes(end);
+    return [startBytes, endBytes];
+  }
+
+  private positionToBytes(position: Position) {
+    // Subtract one as this logic is zero-based and the columns are one-based
+    return this.bytesBeforeLine(position.line) + position.column - 1;
+  }
+
+  private bytesBeforeLine(line: number) {
+    // Subtract one as this logic is zero-based and the lines are one-based
+    line -= 1;
+
+    const precedingLines = this.lines.slice(0, line);
+
+    // Add one to account for the newline we split on and removed
+    return precedingLines.map(l => l.length + 1).reduce((a, b) => a + b);
+  }
+}
+
+const modeId = 'my-rust';
+
+const initMonaco = (monaco: Monaco) => {
+  monaco.editor.defineTheme('vscode-dark-plus', {
```

Comment:
Why do we need to create our own theme? What isn't provided by upstream?

Reply:
VSCode themes are not compatible with monaco because they use textmate convention for tokens but monaco has its own (funnily, monaco is directly built from vscode source code) so upstream won't provide vscode themes. Original themes are ugly (IMO) and people are more familiar with vscode theme.

That theme is not exactly equal to vscode's dark-plus, but it is close enough and I think it is good for default theme (and user can change theme). But we can remove those nine lines if there is something wrong with them.